ironfish/src/rpc/adapters/tcpAdapter.ts

```diff
@@ -0,0 +1,207 @@
+    type: yup.string().oneOf(['message']).required(),
+    data: yup
+      .object({
+        mid: yup.number().required(),
+        type: yup.string().required(),
+        data: yup.mixed().notRequired(),
+      })
+      .required(),
+  })
+  .required()
+
+export class TcpAdapter implements IAdapter {
+  logger: Logger
+  host: string
+  port: number
+  server: net.Server | null = null
+  router: Router | null = null
+  namespaces: ApiNamespace[]
+
+  constructor(
+    host: string,
+    port: number,
+    logger: Logger = createRootLogger(),
+    namespaces: ApiNamespace[],
+  ) {
+    this.host = host
+    this.port = port
+    this.logger = logger.withTag('tcpadapter')
+    this.namespaces = namespaces
+  }
+
+  start(): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.server = net.createServer((socket) => this.onClientConnection(socket))
+
+      this.server.on('error', (err) => {
+        reject(err)
+      })
+
+      this.server.listen(
+        {
+          host: this.host,
+          port: this.port,
+          exclusive: true,
+        },
+        () => {
+          resolve()
+        },
+      )
+    })
+  }
+
+  stop(): Promise<void> {
+    this.logger.debug(`tcpAdapter stopped: ${this.host}:${this.port}`)
+    return Promise.resolve()
+  }
+
+  attach(server: RpcServer): void {
+    this.router = server.getRouter(this.namespaces)
+  }
+
+  unattach(): void {
+    this.router = null
+  }
+
+  onClientConnection(socket: net.Socket): void {
+    socket.on('data', (data) => {
+      this.onClientData(socket, data).catch((err) => this.logger.error(err))
+    })
+    socket.on('close', () => {
+      //TODO: possibly keep track of ongoing reqeusts on this connection and
+      // cancel all of them. This seems like an optimization but not a requirement
+      this.logger.debug(`client connection closed: ${this.host}:${this.port}`)
+    })
+    socket.on('error', (error: Error) => {
+      this.logger.debug(`${this.host}:${this.port} has error : ${error.message}`)
+    })
+  }
+
+  async onClientData(socket: net.Socket, data: Buffer): Promise<void> {
+    const dataString = data.toString('utf8').trim()
+    const result = await YupUtils.tryValidate(IncomingNodeIpcSchema, dataString)
+
+    if (result.error) {
+      this.logger.log(
+        `${this.host}:${this.port} Error parsing message : ${result.error.message} raw: ${dataString} `,
+      )
+    } else {
+      const message = result.result.data
+
+      const request = new Request(
+        message.data,
+        (status: number, data?: unknown) => {
+          const res = this.encodeNodeIpc(this.constructMessage(message.mid, status, data))
+          socket.write(res)
+        },
+        (data: unknown) => {
+          const res = this.encodeNodeIpc(this.constructStream(message.mid, data))
+          socket.write(res)
+        },
+      )
+
+      if (this.router == null) {
+        this.logger.log('handling incoming connection on unmounted adapter')
```

Comment:
same feedback here, because for every error case we have to add an indent, and also double check there's no more code after the else that will run here.

Reply:
Yea that makes sense. Just in case any code gets added on accidentally after 👍 . Fixed